lib/proxy_ssh/proxy.go

```diff
@@ -0,0 +1,282 @@
+	for newChannel := range serverChannels {
+		remoteChannel, remoteRequests, remoteErr := remoteConn.OpenChannel(newChannel.ChannelType(), newChannel.ExtraData())
+		if remoteErr != nil {
+			return log.Errorf("Could not open channel to remote end of proxy: %v", remoteErr)
+		}
+
+		localChannel, localRequests, localErr := newChannel.Accept()
+		if localErr != nil {
+			return log.Errorf("Could not accept server channel: %v", localErr)
+		}
+
+		// Connect the requests.
+		go func() {
+			for {
+				var request *ssh.Request
+				var targetChannel ssh.Channel
+
+				// This is the core "trick" about this infinite for loop: when
+				// data is going back and forth one or both of these get hit
+				// with an actual value.  However, when nothing is going between
+				// this `select` will block unti la request comes in.
+				select {
+				case request = <-localRequests:
+					targetChannel = remoteChannel
+				case request = <-remoteRequests:
+					targetChannel = localChannel
+				}
+
+				// In the event that an SSH request gets killed (not exited),
+				// the request will be nil.  Do not continue, exit the loop.
+				if request == nil {
+					log.Warnf("SSH connection terminated ungracefully...")
+					break
+				}
+
+				requestValid, requestError := targetChannel.SendRequest(request.Type, request.WantReply, request.Payload)
+				if requestError != nil {
+					log.Errorf("SendRequest error: %v", requestError)
+					// TODO: we may need to identify a way to hit this error
+					// in order to decide how to handle it.  For now, we
+					// do nothing other than end the connection.
+					// TODO: probably this is a good time to delete it?
+					p.sessions.Delete(remoteAddrString)
+					break
+				}
+
+				if request.WantReply {
+					request.Reply(requestValid, nil)
+				}
+
+				// NOTE: currently we only care about the exit-status
+				log.Debugf("Request: Type=%q, WantReply='%t'.", request.Type, request.WantReply)
+				if request.Type == "exit-status" {
+					// Connection is closed, end the proxy loop and remove this
+					// session from the list of known sessions.
+					p.sessions.Delete(remoteAddrString)
+					// End the proxy loop
+					break
+				}
+			}
+
+			// End the communication between the local and remote.
+			localChannel.Close()
+			remoteChannel.Close()
+		}()
+
+		// Connect the channels.
+		log.Debugf("Begin session connection between %q and %q.", conn.RemoteAddr(), remoteConn.RemoteAddr())
+		go io.Copy(remoteChannel, localChannel)
+		go io.Copy(localChannel, remoteChannel)
+		// These are kept for safety to ensure the channels are indeed closed,
+		// but in theory the ProxyLoop will close the channels and that will
+		// signal to io.Copy that we are complete.
+		defer localChannel.Close()
+		defer remoteChannel.Close()
+	}
+
+	log.Debugf("Connection between %q and %q closed.", conn.RemoteAddr(), remoteConn.RemoteAddr())
+	return nil
+}
+
+// NOTE: we assume that the `listener` has already been opened and can be used
```

Comment:
Probably outdated comment here from our previous tries to make it work with listener outside of the function.